larq_compute_engine/mlir/python/util.py

```diff
@@ -0,0 +1,201 @@
+            pos, float_tensor, int_tensor = (
+                "input",
+                tensors[op.inputs[0]],
+                tensors[op.outputs[0]],
+            )
+            input_quant_ops.append(op)
+        # Find output dequantize operator
+        elif op.opcodeIndex == dequant_opcode_idx and op.outputs[0] in subgraph.outputs:
+            pos, float_tensor, int_tensor = (
+                "output",
+                tensors[op.outputs[0]],
+                tensors[op.inputs[0]],
+            )
+            output_dequant_ops.append(op)
+        # Otherwise, ignore
+        else:
+            continue
+        # If found, validate the input/output tensor type
+        if float_tensor.type != tflite_schema.TensorType.FLOAT32:
+            raise ValueError(
+                "Model {} type must be tf.float32. Expected type for tensor with "
+                "name '{}' is tf.float32, instead type is tf.{}".format(
+                    pos,
+                    float_tensor.name,
+                    _convert_tflite_enum_type_to_tf_type(float_tensor.type).name,
+                )
+            )
+        if int_tensor.type != tflite_schema.TensorType.INT8:
+            raise ValueError(
+                "Model is not integer quantized. Expected type for tensor with "
+                "name '{}' is tf.int8, instead type is tf.{}".format(
+                    int_tensor.name,
+                    _convert_tflite_enum_type_to_tf_type(int_tensor.type).name,
+                )
+            )
+
+    return input_quant_ops, output_dequant_ops
+
+
+def modify_integer_quantized_model_io_type(
+    model, inference_input_type=tf.float32, inference_output_type=tf.float32,
+):
+    """Modify the float input/output type of an integer quantized model."""
+    # Convert the model to an object
+    model = _convert_model_from_bytearray_to_object(model)
+
+    # Validate the integer quantized model
+    input_quant_ops, output_dequant_ops = _find_int8_quantized_inputs_outputs(model)
+
+    subgraph = model.subgraphs[0]
+    operators = subgraph.operators
+    remove_tensors_idxs = set()
+
+    # Modify model input type
+    if inference_input_type == tf.int8:
+        # Remove the inputs and the quant operator
+        for op in input_quant_ops:
+            subgraph.inputs[subgraph.inputs == op.inputs[0]] = op.outputs[0]
+            remove_tensors_idxs.add(op.inputs[0])
+            operators.remove(op)
+
+    # Modify model output type
+    if inference_output_type == tf.int8:
+        # Remove the outputs and the dequant operator
+        for op in output_dequant_ops:
+            subgraph.outputs[subgraph.outputs == op.outputs[0]] = op.inputs[0]
+            remove_tensors_idxs.add(op.outputs[0])
+            operators.remove(op)
+
+    # Remove tensors marked for deletion.
+    _remove_tensors_from_model(model, remove_tensors_idxs)
+
+    # Convert the model to a bytearray
+    return _convert_model_from_object_to_bytearray(model)
```

Comment:
This is the only function relevant for review in this file and was adapted from [here](https://github.com/tensorflow/tensorflow/commit/ec94fabb838b0934eeaf46f1fd22c36b2b3ec333#diff-84a6aaef794e9a9191371241d009e3bbR685-R780). Main changes are use of native `tf.dtypes` and support for `int8` only.

Reply:
I think it might be nice to keep the [logging line](https://github.com/tensorflow/tensorflow/commit/ec94fabb838b0934eeaf46f1fd22c36b2b3ec333#diff-84a6aaef794e9a9191371241d009e3bbR719-R722) from the commit in TF, for ease of debugging potential Int8 issues. How do you feel about that?

Reply:
I'm not sure if this logging line is relevant for us, since we only support `tf.float32` and `tf.int8` and don't have the complex handling of the different quantization code paths.